src/libraries/System.Console/src/System/Console.cs

```diff
@@ -33,31 +34,54 @@ public static class Console
         private static ConsoleCancelEventHandler? s_cancelCallbacks;
         private static ConsolePal.ControlCHandlerRegistrar? s_registrar;
 
-        internal static T EnsureInitialized<T>([NotNull] ref T? field, Func<T> initializer) where T : class =>
-            LazyInitializer.EnsureInitialized(ref field, ref InternalSyncObject, initializer);
+        public static TextReader In
+        {
+            get
+            {
+                return Volatile.Read(ref s_in) ?? EnsureInitialized();
 
-        public static TextReader In => EnsureInitialized(ref s_in, () => ConsolePal.GetOrCreateReader());
+                static TextReader EnsureInitialized()
+                {
+                    // Must be placed outside s_syncObject lock. See Out getter.
+                    ConsolePal.EnsureConsoleInitialized();
+
+                    lock (s_syncObject) // Ensures In and InputEncoding are synchronized.
+                    {
+                        return s_in ??= ConsolePal.GetOrCreateReader();
```

Comment:
(Technically according to ECMA the write to s_in should be volatile as well, but coreclr doesn't require that (I assume mono, too?) and we don't consistently do it everywhere we lazily initialize.)

Reply:
Out of interest: does ECMA say there is a full fence at the lock exit? Isn't that enough?

Reply:
> Out of interest: does ECMA say there is a full fence at the lock exit? Isn't that enough?

There's a store release on lock exit.  But if the s_in object reference gets published prior to writes that fully construct that published object, that could be visible before even exiting the lock.

Reply:
> the s_in object reference gets published prior to writes that fully construct that published object

Interesting! I had never thought of that.

Should I add `Volatile.Write` where needed for ECMA?

> but coreclr doesn't require that

Does it not require this specifically when constructing new objects? Or more general on writes to memory?
Is this behavior architecture specific?

Reply:
@stephentoub I still have some open questions here.

Reply:
(Sorry, somehow I managed to mute this thread.)

> Should I add Volatile.Write where needed for ECMA?

That would be my preference.  The write needs to be volatile on architectures where it matters; if the runtime does it, then having the Volatile.Write in the code shouldn't hurt (other than making it a bit longer), and if it doesn't, then it's necessary.

coreclr effectively makes every reference assignment into a volatile write for architectures where it matters (e.g. ARM).  @VSadov, @marek-safar, does mono do the same?  If it doesn't, I would not be surprised if our inconsistencies here across code in dotnet/runtime lead to concurrency bugs when running on mono, and we'll either need to a) ensure mono has the same guarantees as coreclr, and/or b) audit all such code in dotnet/runtime to ensure the correct barriers are being used.

Reply:
In Coreclr a managed reference store to heap is a release. You cannot hand off an instance to another thread and not finishing constructing it (unless constructor itself publishes “this”).
Historically that was not the case for Mono.
I believe, we try not rely on this in runtime libraries as a matter of style and use VolatileWrite. 
I am not sure we have no holes though. If we strictly followed the practice everywhere except for guaranteed thread-local writes, I think we would have a lot more volatile in the code. 

As counterexample Roslyn codebase does not do that. In Roslyn compiler most internal data structures are indirectly shared and all threads have the same view. Many writes would have to become Volatile use if Roslyn needed to run on Mono + weak memory.

Reply:
> I am not sure we have no holes though

I'm not a betting man, and I'd bet money we do ;-)

Reply:
VolatileRead is generally necessary only to order between unrelated variables. All current platforms respect indirection as a data dependency. 
Adding VolatileRead when reading a field of an instance is not adding much, unless you want to order with something else, other than the instance.

For unrelated variables write fences generally need to be paired with corresponding read fences.

Reply:
> VolatileRead is generally necessary only to order between unrelated variables.

Interesting. This means the `Volatile.Reads` in this PR aren't needed?

And when a method reads the same reference that is stored on heap multiple times, the JIT/compiler aren't allowed to re-use the values from a local?

Is this ECMA spec? Or coreclr implementation?

Reply:
In .NET Core mode MonoVM follows coreclr memory model.

Reply:
> This means the Volatile.Reads in this PR aren't needed?

It is a good idea to use `Volatile.Read/Write` in lock-free code as a form of documentation.

Reply:
> Interesting. This means the Volatile.Reads in this PR aren't needed?

I did not have a chance to look at the PR in details. 

My general view on this is that pairing read fences with write fences is a good practice. If there was a need for a write fence, it is likely there is a need for a read fence. Unpaired fencing is relatively easy to spot and often require explanation.

Extra read fences do not hurt much. They may interfere with code clarity or become full fences on ARM32, although the platform has been a bit on the downhill lately.

By default, **I'd keep read/write  fence symmetry, or add comments why not needed.**

Having said that, I'd support a policy change for not using fences for publication purpose. That is the case where we already have a paired fence. (including Mono, it seems)

Having too many fences could be a distraction from the cases where fencing is actually needed and requires extraordinary care.

In cases where fencing is actually needed, it could be a good practice to add a comment "need to write Y after writing X".
From my experience - The person who reads or makes changes to the code can see that you are ordering the write of Y. But ordering _relative to what_?
Even if that person is you, just a few weeks later, you may find it valuable to have a comment about X. When a read fence is paired symmetrically, having the comment just on the write side could be enough.

Reply:
Thanks for explaining!

I'll add the `Volatile.Write` and keep the corresponding `Volatile.Read` in place.